src/analyses/ai.cpp

```diff
@@ -239,126 +240,159 @@ bool ai_baset::fixedpoint(
+      goto_program.get_successors(l).empty(),
+      "The end function instruction should have no successors.");
 
-    std::unique_ptr<statet> tmp_state(
-      make_temporary_state(current));
+    // Do nothing
+  }
+  else
+  {
+    // Successors can be empty, for example assume(0).
+    // Successors can contain duplicates, for example GOTO next;
+    for(const auto &to_l : goto_program.get_successors(l))
+    {
+      if(to_l == goto_program.instructions.end())
+        continue;
 
-    statet &new_values=*tmp_state;
+      new_data |= visit_edge(h, working_set, to_l, ns);
+    }
+  }
 
-    bool have_new_values=false;
+  return new_data;
+}
 
-    if(l->is_function_call() &&
-       !goto_functions.function_map.empty())
-    {
-      // this is a big special case
-      const code_function_callt &code=
-        to_code_function_call(l->code);
-
-      if(do_function_call_rec(
-          l, to_l,
-          code.function(),
-          code.arguments(),
-          goto_functions, ns))
-        have_new_values=true;
-    }
-    else
-    {
-      // initialize state, if necessary
-      get_state(to_l);
+bool ai_baset::visit_edge(
+  const tracet &h,
+  working_sett &working_set,
+  const locationt &to_l,
+  const namespacet &ns)
+{
+  // Has history taught us not to step here...
+  auto next = step(h, to_l);
+  if(next.first == ai_history_baset::step_statust::BLOCKED)
+    return false;
+  const tracet &to_h = *(next.second);
 
-      new_values.transform(l, to_l, *this, ns);
+  // Abstract domains are mutable so we must copy before we transform
+  statet &current = get_state(h);
 
-      if(merge(new_values, l, to_l))
-        have_new_values=true;
-    }
+  std::unique_ptr<statet> tmp_state(make_temporary_state(current));
+  statet &new_values = *tmp_state;
 
-    if(have_new_values)
-    {
-      new_data=true;
-      put_in_working_set(working_set, to_l);
-    }
+  // Apply transformer
+  new_values.transform(
+    h.current_location(), to_h.current_location(), *this, ns);
+
+  // Initialize state(s), if necessary
+  get_state(to_h);
+
+  if(
+    merge(new_values, h, to_h) ||
+    next.first == ai_history_baset::step_statust::NEW_FORCE_CONTINUE)
+  {
+    put_in_working_set(working_set, next.second);
+    return true;
   }
 
-  return new_data;
+  return false;
 }
 
+/// Remember that h_call and h_return are both in the caller
 bool ai_baset::do_function_call(
-  locationt l_call, locationt l_return,
+  const tracet &h_call,
   const goto_functionst &goto_functions,
   const goto_functionst::function_mapt::const_iterator f_it,
   const exprt::operandst &arguments,
   const namespacet &ns)
 {
-  // initialize state, if necessary
-  get_state(l_return);
-
+  locationt l_call = h_call.current_location();
   PRECONDITION(l_call->is_function_call());
 
   const goto_functionst::goto_functiont &goto_function=
     f_it->second;
 
+  locationt l_return = std::next(l_call);
+  // DATA_INVARIANT(l_return.is_dereferenceable(),
```

Comment:
Could make this real like the != end() check below?

Reply:
I don't think so because you don't have the caller goto_program object and so have nothing to ```.end()```.  (Another instance of the "maybe we shouldn't use iterators to describe locations" ?)

Reply:
Ah apologies, I didn't realise that was the callee